### PR TITLE
Causal trees update

### PR DESCRIPTION
**Pull request by alexander-pv**
## Proposed changes

This PR is mainly about causal trees support.

* The architecture of a causal tree implementation was moved to a more modular approach:

1.  `BaseCausalDecisionTree` inherits everything is needed from scikit-learn `BaseDecisionTree` and modifies `fit()` method that stores only appropriate checks for causal trees.
2. `CausalTreeRegressor` now has `RegressorMixin` and `BaseCausalDecisionTree` parent classes which makes it fully compatible with scikit-learn.
3. Split criterion was moved to a separate `criterion.pyx` where `CausalRegressionCriterion` inherits methods from scikit-learn `RegressionCriterion` and implements `node_value()` to save the average of treatment effect for each node. `CausalMSE` now is a concrete class with impurity computations for causal trees. I also added `StandardMSE` concrete class which is actually standard MSE criterion from scikit-learn with modified `node_value()` method. So, now it is easy to add new criteria and see the influence of each criteria on a causal tree fit .

* Details about causal trees:

1. ATE bootstrap confidence intervals calculation in `CausalTreeRegressor`  now has multiprocessing support.
2. Now you can plot `CausalTreeRegressor` with standard scikit-learn function.
3. For a deeper research `CausalTreeRegressor` can calculate the number of treatment and control observations in each leaf, `_leaves_groups_cnt` low-level attribute. Additionally, `plot_dist_tree_leaves_values` function gives the distribution of ATE in a tree leaves.
4. `CausalRandomForestRegressor` based on scikit-learn with `CausalTreeRegressor` as `base_estimator`.
5. Method `calculate_error`  in `CausalRandomForestRegressor` calculates unbiased sampling variance. [Source](https://github.com/scikit-learn-contrib/forest-confidence-interval).
6. New Jupyter notebook `causal_trees_with_synthetic_data.ipynb` with `CausalTreeRegressor` and `CausalRandomForestRegressor` models.

* Tests:
1. Additional tests: `test_causal_trees.py`
2. `Makefile` contains install, build, test, clean. Now you can simply type `make test`. Cython code compilation is under the hood.
3. `setup()` function in `setup.py` now knows about `requirements-test.txt` dependencies thanks to `tests_require` parameter. No need to install them manually.

## Types of changes

What types of changes does your code introduce to **CausalML**?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/uber/causalml/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/uber/causalml)
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc. This PR template is adopted from [appium](https://github.com/appium/appium).

### CODE REVIEW

**Comment by jeongyoonlee** [comment]
Thanks @alexander-pv for your much needed contribution. We will review the PR soon.

File: causalml/inference/tree/causal/causaltree.py [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+from pathos.pools import ProcessPool as PPool
+from scipy.stats import norm
+from sklearn.base import RegressorMixin
+from sklearn.ensemble._forest import ForestRegressor
+from sklearn.tree._tree import DTYPE
+from sklearn.utils import check_array
+from sklearn.utils.validation import check_is_fitted
+
+from causalml.inference.meta.utils import check_treatment_vector
+from .base_tree import BaseCausalDecisionTree
```
Comment (jeongyoonlee) [inline causalml/inference/tree/causal/causaltree.py:16]: `.base_tree` --> `._tree`

File: causalml/inference/tree/causal/causaltree.py [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+                - If "sqrt", then `max_features=sqrt(n_features)`.
+                - If "log2", then `max_features=log2(n_features)`.
+                - If None, then `max_features=n_features`.
+            max_leaf_nodes: (int, default=None)
+                Grow a tree with ``max_leaf_nodes`` in best-first fashion.
+                Best nodes are defined as relative reduction in impurity.
+                If None then unlimited number of leaf nodes.
+            min_impurity_split : (float, default=float("-inf"))
+                Threshold for early stopping in tree growth. A node will split
+                if its impurity is above the threshold, otherwise it is a leaf.
```
Comment (jeongyoonlee) [inline causalml/inference/tree/causal/causaltree.py:90]: Let's remove it as mentioned above.

File: causalml/inference/tree/causal/causaltree.py [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+        self.splitter = splitter
+        self.alpha = alpha
+        self.control_name = control_name
+        self.max_depth = max_depth
+        self.min_samples_split = min_samples_split
+        self.min_weight_fraction_leaf = min_weight_fraction_leaf
+        self.max_features = max_features
+        self.max_leaf_nodes = max_leaf_nodes
+        self.min_impurity_decrease = min_impurity_decrease
+        self.min_impurity_split = min_impurity_split
```
Comment (jeongyoonlee) [inline causalml/inference/tree/causal/causaltree.py:126]: Same as above

File: causalml/inference/tree/causal/causaltree.py [hunk 4]
```diff
@@ -0,0 +1,615 @@
...
+        super().__init__(
+            criterion=criterion,
+            splitter=splitter,
+            max_depth=max_depth,
+            min_samples_split=min_samples_split,
+            min_weight_fraction_leaf=min_weight_fraction_leaf,
+            max_features=max_features,
+            max_leaf_nodes=max_leaf_nodes,
+            min_impurity_decrease=min_impurity_decrease,
+            min_impurity_split=min_impurity_split,
```
Comment (jeongyoonlee) [inline causalml/inference/tree/causal/causaltree.py:145]: same as above

File: causalml/inference/tree/causal/causaltree.py [hunk 5]
```diff
@@ -0,0 +1,615 @@
...
+            treatment: : (np.ndarray), treatment vector
+            sample_weight: (np.ndarray), sample_weight
+            check_input: (bool)
+        Returns:
+            self
+        """
+
+        if (
+            self.criterion == "causal_mse"
+            and self.min_impurity_split != float("-inf")
```
Comment (jeongyoonlee) [inline causalml/inference/tree/causal/causaltree.py:173]: same as above

File: causalml/inference/tree/causal/causaltree.py [hunk 6]
```diff
@@ -0,0 +1,615 @@
...
+        control_name: Union[int, str] = 0,
+        criterion: str = "causal_mse",
+        alpha: float = 0.05,
+        max_depth: int = None,
+        min_samples_split: int = 2,
+        min_samples_leaf: int = 100,
+        min_weight_fraction_leaf: float = 0.0,
+        max_features: Union[int, float, str] = 1.0,
+        max_leaf_nodes: int = None,
+        min_impurity_split: float = float("-inf"),
```
Comment (jeongyoonlee) [inline causalml/inference/tree/causal/causaltree.py:447]: same as above

File: causalml/inference/tree/causal/causaltree.py [hunk 7]
```diff
@@ -0,0 +1,615 @@
...
+                The minimum number of samples required to be at a leaf node.
+            min_weight_fraction_leaf: (float, default=0.0)
+                The minimum weighted fraction of the sum total of weights (of all
+                the input samples) required to be at a leaf node.
+            max_features: (int, float or {"auto", "sqrt", "log2"}, default=None)
+                The number of features to consider when looking for the best split
+            max_leaf_nodes: (int, default=None)
+                Grow a tree with ``max_leaf_nodes`` in best-first fashion.
+            min_impurity_split : (float, default=float("-inf"))
+                Threshold for early stopping in tree growth.
```
Comment (jeongyoonlee) [inline causalml/inference/tree/causal/causaltree.py:484]: same as above

File: causalml/inference/tree/causal/causaltree.py [hunk 8]
```diff
@@ -0,0 +1,615 @@
...
+            n_estimators=n_estimators,
+            estimator_params=(
+                "criterion",
+                "control_name",
+                "max_depth",
+                "min_samples_split",
+                "min_weight_fraction_leaf",
+                "max_features",
+                "max_leaf_nodes",
+                "min_impurity_split",
```
Comment (jeongyoonlee) [inline causalml/inference/tree/causal/causaltree.py:524]: same as above

File: causalml/inference/tree/causal/causaltree.py [hunk 9]
```diff
@@ -0,0 +1,615 @@
...
+
+        self.criterion = criterion
+        self.control_name = control_name
+        self.max_depth = max_depth
+        self.min_samples_split = min_samples_split
+        self.min_samples_leaf = min_samples_leaf
+        self.min_weight_fraction_leaf = min_weight_fraction_leaf
+        self.max_features = max_features
+        self.max_leaf_nodes = max_leaf_nodes
+        self.min_impurity_split = min_impurity_split
```
Comment (jeongyoonlee) [inline causalml/inference/tree/causal/causaltree.py:547]: same as above

File: causalml/inference/tree/causal/_tree.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+        # Use BestFirst if max_leaf_nodes given; use DepthFirst otherwise
+        if max_leaf_nodes < 0:
+            builder = DepthFirstTreeBuilder(
+                splitter,
+                min_samples_split,
+                min_samples_leaf,
+                min_weight_leaf,
+                max_depth,
+                self.min_impurity_decrease,
+                min_impurity_split,
```
Comment (jeongyoonlee) [inline causalml/inference/tree/causal/_tree.py:230]: Let's remove it

File: causalml/inference/tree/causal/_tree.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+        else:
+            builder = BestFirstTreeBuilder(
+                splitter,
+                min_samples_split,
+                min_samples_leaf,
+                min_weight_leaf,
+                max_depth,
+                max_leaf_nodes,
+                self.min_impurity_decrease,
+                min_impurity_split,
```
Comment (jeongyoonlee) [inline causalml/inference/tree/causal/_tree.py:241]: Let's remove it

File: causalml/inference/tree/causal/_tree.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+        if sample_weight is None:
+            min_weight_leaf = self.min_weight_fraction_leaf * n_samples
+        else:
+            min_weight_leaf = self.min_weight_fraction_leaf * np.sum(sample_weight)
+
+        min_impurity_split = self.min_impurity_split
+
+        # min_impurity_split is deprecated
+        if min_impurity_split is None:
+            min_impurity_split = 0
```
Comment (jeongyoonlee) [inline causalml/inference/tree/causal/_tree.py:184]: Let's remove min_impurity_split to avoid errors with latest sklearn

**Review by jeongyoonlee — CHANGES_REQUESTED** [review]

**Comment by jeongyoonlee** [comment]
Hi, @alexander-pv, 

Sorry for my late review. I thought that I left comments earlier but turned out that I didn't submit them and my review was somehow left pending. Code looks good and thanks for the sample notebook and test code which are very comprehensive. One ask is: let's remove `min_impurity_split` from the code so that we can use latest scikit-learn. Currently it raises "TypeError: __init__() got an unexpected keyword argument 'min_impurity_split'".

Thanks!

**Comment by alexander-pv** [comment]
Hi, @jeongyoonlee ,

Thanks for the review, I pushed necessary changes.

I faced the thing that `min_impurity_split` removal in latest scikit-learn breaks tree building.  It turned out that new version of `DepthFirstTreeBuilder` prevent causal tree from growing new nodes.
See differences: [old condition](https://github.com/scikit-learn/scikit-learn/blob/2beed55847ee70d363bdbfe14ee4401438fba057/sklearn/tree/_tree.pyx#L221) vs [new condition](https://github.com/scikit-learn/scikit-learn/blob/baf828ca126bcb2c0ad813226963621cafe38adb/sklearn/tree/_tree.pyx#L216).
Since `EPSILON` constant in original cython file is widely used and can't be adjusted, I suggest to create `builder.pyx` with DepthFirstCausalTreeBuilder with necessary changes (see latest commits). This file can also be an example how to create custom tree builders via subclassing scikit-learn TreeBuilder.

**Review by jeongyoonlee — APPROVED** [review]
LGTM. Thanks again for your contribution. I really appreciate it!

**Comment by jeongyoonlee** [comment]
Hi @alexander-pv, Py 3.7 test failed with the error as follows:
https://github.com/uber/causalml/runs/7923168250?check_suite_focus=true#step:6:492

Could you please check? If it's something that needs more time, we can merge this PR first and investigate it in a separate thread.

**Comment by alexander-pv** [comment]
Hi, @jeongyoonlee, It seems that now everything is fixed :ok_hand:.

**Comment by jeongyoonlee** [comment]
Merged! 👍🏻

**Pull request merged**